packages/dds/tree/src/shared-tree/schematizedTree.ts

```diff
@@ -0,0 +1,124 @@
+ * The returned view referees to and can edit the provided one: it is not a fork of it.
+ * Updates the stored schema in the tree to match the provided one if requested by config and compatible.
+ *
+ * If the tree is uninitialized (has no nodes or schema at all),
+ * it is initialized to the config's initial tree and the provided schema are stored.
+ * This is done even if `AllowedUpdateType.None`.
+ *
+ * TODO:
+ * - Implement schema-aware API for return type.
+ * - Support adapters for handling out of schema data.
+ * - Support lazy schema updates.
+ * - Improve discoverability of this by either integrating it into the factory or main SharedTree interfaces as a method.
+ * @alpha
+ */
+export function schematizeBranch(
+	tree: ISharedTreeBranch,
+	config: SchematizeConfiguration,
+): ISharedTreeBranch {
+	// Check for empty.
+	// This case could be better handled by making a schematized factory (so it runs only in the "create" case),
+	// but handling it here works well enough, and can be considered a special case of an out of schema handler.
+	// Note that if this becomes a more proper out of schema handler, it should be made lazy.
+	{
+		if (tree.context.root.length === 0 && schemaDataIsEmpty(tree.storedSchema)) {
+			tree.storedSchema.update(config.schema);
+			tree.root = config.initialTree;
+		}
+	}
+
+	// TODO: support adapters and include them here.
+	const viewSchema = new ViewSchema(defaultSchemaPolicy, {}, config.schema);
+	{
+		const compatibility = viewSchema.checkCompatibility(tree.storedSchema);
+		switch (config.allowedSchemaModifications) {
+			case AllowedUpdateType.None: {
+				if (compatibility.read !== Compatibility.Compatible) {
+					fail(
+						"Existing stored schema permits trees which are incompatible with the view schema",
+					);
+				}
+
+				if (compatibility.write !== Compatibility.Compatible) {
+					// TODO: support readonly mode in this case.
+					fail("View schema permits trees which are incompatible with the stored schema");
+				}
+
+				break;
+			}
+			case AllowedUpdateType.SchemaCompatible: {
+				if (compatibility.read !== Compatibility.Compatible) {
+					fail(
+						"Existing stored schema permits trees which are incompatible with the view schema, so schema can not be updated",
+					);
+				}
+				if (compatibility.write !== Compatibility.Compatible) {
+					tree.storedSchema.update(config.schema);
+				}
+
+				break;
+			}
+			default: {
+				unreachableCase(config.allowedSchemaModifications);
+			}
+		}
+	}
+
+	// TODO: errors thrown by this will usually be in response to remote edits, and thus may not surface to the app.
+	// Two fixes should be done related to this:
+	// 1. Ensure errors in response to edits like this crash app and report telemetry.
+	// 2. Replace these (and the above) exception based errors with
+	// out of schema handlers which update the schematized view of the tree instead of throwing.
+	tree.storedSchema.registerDependent(
+		new SimpleObservingDependent(() => {
```

Comment:
Can the same tree ever be schematized twice? If so, we will register this dependent multiple times; do we care about unregistering this ever? (Probably not, if we're trying to crash the app, but just checking)

Reply:
if you did, you would want to keep this event anyway because you could still have code accessing the tree using the first view schema types.


I don't recommend doing this since we don't currently have any support of out of schema adapters either the two views would be the same, or one would error, but in the future when schematize can actually apply adapters and doesn't just return its input, you could do it and it could be useful (though I don't recommend it as a pattern).

Once we have a schema aware API, passing the output of schematize back into schematize won't typecheck (unless we want it to), but you could still have two different schematized projections of the same (or different) views.